pkg/tsdb/phlare/query.go

```diff
@@ -179,111 +179,118 @@ func (pt ProfileTree) String() string {
 	return tree.String()
 }
 
-// merge merges the node into the tree.
-// it assumes src has only one leaf.
-func (pt *ProfileTree) merge(src *ProfileTree) {
-	// find the node path where n should be inserted.
-	var parent, found *ProfileTree
-	// visit depth first the dst tree following the src tree
-	remaining := []*ProfileTree{pt}
-	for len(remaining) > 0 {
-		n := remaining[0]
-		remaining = remaining[1:]
-		if src.locationID == n.locationID {
-			if len(src.Nodes) == 0 {
-				// we have found the leaf
-				found = n
-				break
-			}
-			// move src and last parent visited
-			parent = n
-			src = src.Nodes[0]
-			remaining = n.Nodes
-			continue
-		}
+// addSample adds a sample to the tree. As sample is just a single stack we just have to traverse the tree until it
+// starts to differ from the sample and add a new branch if needed. For example if we have a tree:
+//
+//	root --> func1 -> func2 -> func3
+//		 \-> func4
+//
+// And we add a sample:
+//
+//	func1 -> func2 -> func5
+//
+// We will get:
+//
+//	root --> func1 --> func2 --> func3
+//	     \                   \-> func5
+//	      \-> func4
+//
+// While we add the current sample value to root -> func1 -> func2.
+func (pt *ProfileTree) addSample(profile *googlev1.Profile, sample *googlev1.Sample) {
```

Comment:
It seemed easier to work with single tree and single sample (just list of IDs) which means the iteration can be simple `for range` + tree traversal instead of 2 tree traversals. It also means we don't have to create a tree in cases where sample just adds values to existing nodes and does not create new nodes.

Reply:
Make total sense to me.